library/core/src/cell.rs

```diff
@@ -388,8 +388,9 @@ impl<T> Cell<T> {
         // is `!Sync` so this won't happen. This also won't invalidate any
         // pointers since `Cell` makes sure nothing else will be pointing into
         // either of these `Cell`s.
+        // We checked above that we don't point to same location.
         unsafe {
-            ptr::swap(self.value.get(), other.value.get());
+            ptr::swap_nonoverlapping(self.value.get(), other.value.get(), 1);
```

Comment:
I'm not sure this is correct.  You checked that they don't point to the *same* location, but not that they're non-overlapping.

For example, MIRI doesn't complain about this today: https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=f540276c645f039e91df3d70233fce6b

Now, I don't know what the correct behaviour for that even is, because swapping things that overlap is super confusing to me.  But I don't know that it should become UB, which I think this change would do.

(The UCG or libs-api folks might have opinions here about what the right thing to do it.)